### PR TITLE
refactor: improve performance for cmo document conversion and sealing (FPLA-2056)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2056

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by Jamesnelson117** [comment]
Ignore additional commits. My local master needs reset by the looks of it.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSealingService.java [hunk 1]
```diff
@@ -32,12 +35,22 @@
 
     private final DocumentDownloadService documentDownloadService;
     private final UploadDocumentService uploadDocumentService;
+    private final DocumentConversionService documentConversionService;
 
-    public DocumentReference sealDocument(DocumentReference document) throws Exception {
-        byte[] documentContent = documentDownloadService.downloadDocument(document.getBinaryUrl());
-        byte[] sealedDocument = sealDocument(documentContent);
+    public DocumentReference sealAndUploadDocument(DocumentReference document) throws Exception {
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocumentSealingService.java:40]: i think sealDocument(Document reference) is better. 
as an input user gives link to remote document, so expectation is that remote document will be sealed, no need to express it in name, client of this method does not care if document is download and then uploaded. Client cares only that remote doc is sealed

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java [hunk 2]
```diff
@@ -87,11 +83,9 @@ void shouldSetReturnStatusAndRequestedChangesWhenJudgeRejectsOrder() {
 
     @Test
     void shouldSealPDFAndAddToSealedCMOsListWhenJudgeApprovesOrder() throws Exception {
-        DocumentReference convertedDocument = testDocumentReference();
         DocumentReference sealedDocument = testDocumentReference();
 
-        given(documentConversionService.convertToPdf(cmo.getOrder())).willReturn(convertedDocument);
-        given(documentSealingService.sealDocument(convertedDocument)).willReturn(sealedDocument);
+        given(documentSealingService.sealAndUploadDocument((cmo.getOrder()))).willReturn(sealedDocument);
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:88]: you have unused variables at class level

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/DocumentSealingServiceTest.java [hunk 3]
```diff
@@ -32,28 +34,59 @@
...
 
+    @Mock
+    private DocumentConversionService documentConversionService;
+
     @InjectMocks
     private DocumentSealingService documentSealingService;
 
     @Test
-    void shouldSealDocument() throws Exception {
+    void shouldSealAndUploadDocumentWithFileExtensionConversion() throws Exception {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/DocumentSealingServiceTest.java:44]: withFileConversion probably

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/DocumentReference.java [hunk 4]
```diff
@@ -28,4 +30,8 @@ public static DocumentReference buildFromDocument(Document document) {
     public boolean isEmpty() {
         return url == null && filename == null && binaryUrl == null;
     }
+
+    public boolean hasExtensionTypeOf(DocumentExtension documentExtension) {
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/DocumentReference.java:34]: Yes we can revert and use `FilenameUtils.isExtension(document.getFilename(), "pdf)` as suggested above

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/DocumentConversionService.java [hunk 5]
```diff
@@ -14,39 +13,18 @@
...
-
-            byte[] documentContent = documentDownloadService.downloadDocument(document.getBinaryUrl());
-            byte[] convertedDocument = convertDocument(documentContent, oldName, newName);
-
-            return buildFromDocument(uploadDocumentService.uploadPDF(convertedDocument, newName));
-        }
-    }
-
-    private byte[] convertDocument(byte[] binaries, String oldName, String newName) {
+    public byte[] convertDocument(byte[] binaries, String oldName, String newName) {
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/docmosis/DocumentConversionService.java:27]: Correct me if i'm wrong but this method is more generic than just pdf conversion no?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/DocumentReferenceTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package uk.gov.hmcts.reform.fpl.model;
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/DocumentReferenceTest.java:1]: there is already DocumentReferenceTest in other package, please merge them

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java [hunk 7]
```diff
@@ -119,7 +111,6 @@ void shouldSealPDFAndAddToSealedCMOsListWhenJudgeApprovesOrder() throws Exceptio
     @Test
     void shouldUpdateStateToIssueResolutionWhenNextHearingTypeIsIssueResolutionAndCmoDecisionIsSendToAllParties()
```
Comment (tomaszpowroznik) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:112]: these 2 test are identical except HearingType, can be parametrised

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/sdo/StandardDirectionsOrderServiceTest.java [hunk 8]
```diff
@@ -128,12 +128,11 @@ void shouldNotSealDocumentWhenSDOIsDraft() throws Exception {
 
         assertThat(builtOrder).isEqualTo(expectedOrder);
         verify(sealingService, never()).sealDocument(any(DocumentReference.class));
-        verify(conversionService, never()).convertToPdf(any(DocumentReference.class));
+        verify(conversionService, never()).convertToPdf(any(), any());
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/sdo/StandardDirectionsOrderServiceTest.java:131]: these 2 assertions can be replaced with
  verifyNoInteractions(sealingService, conversionService);

File: service/src/test/java/uk/gov/hmcts/reform/fpl/utils/TestDataHelper.java [hunk 9]
```diff
@@ -48,6 +48,10 @@ public static DocumentReference testDocumentReference() {
             .build();
     }
 
+    public static DocumentReference testDocumentReference(String fileName) {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/utils/TestDataHelper.java:51]: maybe:

  public static DocumentReference testDocumentReference(String filename) {
        return DocumentReference.builder()
            .filename(filename)
            .url(randomAlphanumeric(10))
            .binaryUrl(randomAlphanumeric(10))
            .build();
    }

    public static DocumentReference testDocumentReference() {
        return testDocumentReference(randomAlphanumeric(10));
    }

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/DocumentsHelper.java [hunk 10]
```diff
@@ -40,4 +42,19 @@ public static String concatUrlAndMostRecentUploadedDocumentPath(String url,
         }
         return "";
     }
+
+    public static boolean hasExtension(DocumentReference document, String extension) {
+        return hasExtension(document.getFilename(), extension);
+    }
+
+    public static boolean hasExtension(String fileName, String extension) {
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/utils/DocumentsHelper.java:50]: filename

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java (hunk 7 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ReviewCMOControllerAboutToSubmitTest.java:112]: Discussed to leave.

**Review by tomaszpowroznik — APPROVED** [review]

**Pull request merged**